File: fab-core/src/main/java/eu/stratosphere/fab/core/beans/experiment/ExperimentSuite.java
package eu.stratosphere.fab.core.beans.experiment;

import eu.stratosphere.fab.core.beans.ExecutionContext;

import java.util.List;

/**
 * Created by alexander on 27.05.14.
 */
public class ExperimentSuite {

    public final List<eu.stratosphere.fab.core.beans.experiment.Experiment> experiments;

    public ExperimentSuite(List<eu.stratosphere.fab.core.beans.experiment.Experiment> experiments) {
        this.experiments = experiments;
    }

    public void run() {
        // TODO: add logic here

        // 1.a) construct the systems graph (this can be done at the bottom of the constructor as well)
        // 1.b) create an execution context object referencing the systems dependency graph
        ExecutionContext context = new ExecutionContext();

        // run the experiments
        // 3.a) setup all systems with 'suite' lifespan in the execution context

        // 3.b) run each experiment in the experiments list
        for (Experiment e: experiments) {
            e.run(context);
        }

        // 3.c) tear down all systems with 'suite' lifespan in the execution context
    }
}

File: fab-core/src/main/scala/eu/stratosphere/fab/core/beans/experiment/ExperimentSuite.scala
package eu.stratosphere.fab.core.beans.experiment

import eu.stratosphere.fab.core.{ExecutionContext, Node, DependencyGraph}
import eu.stratosphere.fab.core.beans.system.{Lifespan, ExperimentRunner, System}
import org.slf4j.LoggerFactory

/**
 * Created by felix on 02.06.14.
 */
class ExperimentSuite(final val experiments: List[Experiment]) extends Node{

  final val logger = LoggerFactory.getLogger(this.getClass)

  def run() =  {

    val depGraph: DependencyGraph[Node] = createGraph()
    val context: ExecutionContext = new ExecutionContext(depGraph)


    //TODO check for cycles in the graph
    if(!depGraph.isEmpty) {
      logger.info("Succesfulyy created experiment graph: \n " + depGraph)
    }
    else {
      throw new RuntimeException("Could not create Graph! (Graph is empty)")
    }

    // setup all systems with suite lifecycle
    setUpSuite(depGraph)
    for(e <- experiments) new ExperimentSequence(e).run(context)

    tearDownSuite(depGraph)

  }

  /**
   * Set up all systems with lifecycle suite in the graph
   * @param g the dependency graph
   * @return nothing
   */
  def setUpSuite(g: DependencyGraph[Node]) = {
    for(s <- g.reverse.dfs()) yield {s match {
      case s: System => if (s.lifespan == Lifespan.SUITE) s.setUp()
      case x => x
    }}
  }

  def tearDownSuite(g: DependencyGraph[Node]) = {
    for(s <- g.reverse.dfs()) yield {s match {
      case s: System => if (s.lifespan == Lifespan.SUITE) s.tearDown()
      case x => x
    }}
  }



  /**
   * create a directed Graph from all Experiments and their dependencies
   * @return Graph with systems as vertices and dependencies as edges
   */
  def createGraph(): DependencyGraph[Node] = {
    val g = new DependencyGraph[Node]

    def getDependencies(s: System): Unit = {
      if(!s.dependencies.isEmpty) {
        for(d <- s.dependencies) yield {
          g.addEdge(s, d)
          getDependencies(d)
        }
      }
    }

    for(e <- experiments) yield {
      g.addEdge(this, e)
      val r: ExperimentRunner = e.runner
      g.addEdge(e, r)
      getDependencies(r)
    }

    g // return Graph
  }


}
